api.js

```diff
@@ -191,44 +217,57 @@ Api.prototype.run = function (files) {
 			self.fileCount = files.length;
 			self.base = path.relative('.', commonPathPrefix(files)) + path.sep;
 
-			var tests = files.map(self._runFile);
-
-			// receive test count from all files and then run the tests
-			var statsCount = 0;
-
 			return new Promise(function (resolve) {
-				tests.forEach(function (test) {
-					function tryRun() {
-						if (++statsCount === self.fileCount) {
-							self.emit('ready');
-
-							var method = self.options.serial ? 'mapSeries' : 'map';
-							var options = {
-								runOnlyExclusive: self.hasExclusive
-							};
-
-							resolve(Promise[method](files, function (file, index) {
-								return tests[index].run(options).catch(function (err) {
-									// The test failed catastrophically. Flag it up as an
-									// exception, then return an empty result. Other tests may
-									// continue to run.
-									self._handleExceptions({
-										exception: err,
-										file: file
-									});
-
-									return {
-										stats: {passCount: 0, skipCount: 0, todoCount: 0, failCount: 0},
-										tests: []
-									};
+				// receive test count from all files and then run the tests
+				var statsCount = 0;
+
+				var tests = new Array(files.length);
+				files.forEach(function (file, index) {
+					self._runFile(file, function (forkManager) {
+						tests[index] = forkManager;
+						forkManager.on('stats', tryRun);
+					})
+					.catch(function (error) {
```

Comment:
This is asynchronous whereas `onForkStarting` is not. Probably best to build up `tests` similarly (and synchronously) to how it was done before and leave the rest of the code as it is. (Note that I just changed this you so you'll need to rebase).